### PR TITLE
Rewrite core of PyMC3 (upcoming v4) to rely on aesara (among many other improvements)

### PR DESCRIPTION
**Pull request by twiecki**

### CODE REVIEW

File: pymc3/blocking.py [hunk 1]
```diff
@@ -18,245 +18,80 @@
...
 
-VarMap = collections.namedtuple("VarMap", "var, slc, shp, dtyp")
-DataMap = collections.namedtuple("DataMap", "list_ind, slc, shp, dtype, name")
+# `point_map_info` is a tuple of tuples containing `(name, shape, dtype)` for
+# each of the raveled variables.
+RaveledVars = collections.namedtuple("RaveledVars", "data, point_map_info")
 
 
-# TODO Classes and methods need to be fully documented.
+class DictToArrayBijection:
```
Comment (michaelosthege) [inline pymc3/blocking.py:38]: Well condensed!

File: pymc3/distributions/__init__.py [hunk 2]
```diff
@@ -11,6 +11,373 @@
...
+from aesara.tensor.var import TensorVariable
+
+from pymc3.aesaraf import floatX
+
+PotentialShapeType = Union[
+    int, np.ndarray, Tuple[Union[int, Variable], ...], List[Union[int, Variable]], Variable
+]
+
+
+def _get_scaling(total_size, shape, ndim):
```
Comment (michaelosthege) [inline pymc3/distributions/__init__.py]: Adding the type hints wouldn't hurt - they are already in the docstring

File: pymc3/distributions/__init__.py [hunk 3]
```diff
@@ -11,6 +11,373 @@
...
+        shp_begin = shape[: len(begin)]
+        begin_coef = [floatX(t) / shp_begin[i] for i, t in enumerate(begin) if t is not None]
+        end_coef = [floatX(t) / shp_end[i] for i, t in enumerate(end) if t is not None]
+        coefs = begin_coef + end_coef
+        coef = aet.prod(coefs)
+    else:
+        raise TypeError(
+            "Unrecognized `total_size` type, expected int or list of ints, got %r" % total_size
+        )
+    return aet.as_tensor(floatX(coef))
```
Comment (michaelosthege) [inline pymc3/distributions/__init__.py]: If I'm not mistaken the `floatX` here is not necessary

**Comment by ricardoV94** [comment]
In case this helps:

1. Since https://github.com/pymc-devs/pymc3/commit/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d, prior sampling returns the same value for every draw.

```python
with pm.Model() as m:
    x = pm.Normal('x', 0, 1)
    prior = pm.sample_prior_predictive(10)
    print(prior)
    
{'x': array([0.019066, 0.019066, 0.019066, 0.019066, 0.019066, 0.019066,
       0.019066, 0.019066, 0.019066, 0.019066])}
```

2. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?

**Comment by brandonwillard** [comment]
> 1. Since [a80cf9a](https://github.com/pymc-devs/pymc3/commit/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d), prior sampling returns the same value for every draw.

Yes, that's because it's using the same RNG seed every time.  The reason why: `RandomVariable.inplace == False`.  [This code](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/distribution.py#L87) was supposed to take care of that.

**Comment by brandonwillard** [comment]
> 1. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?

Is there an example of this, or a test that's failing?

**Comment by ricardoV94** [comment]
> > 1. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?
> 
> Is there an example of this, or a test that's failing?

The random tests are still disabled, so this wouldn't show up right? I wanted to check with the prior predictive sampling but right now that's also unusable. 

There was an issue in the logp of the HalfCauchy that lead to the addition of an unused parameter (not sure if related):
https://github.com/pymc-devs/pymc3/pull/4508/files/3db730e1b632e0fa023cd181c60d19de4aab6e48#r594550615

**Comment by ricardoV94** [comment]
Managed to do it with `.dist`

```python
import pymc3 as pm
import scipy.stats as st

pymc_samples = pm.HalfCauchy.dist(beta=.01, size=10_000).eval()
print(pymc_samples.mean(), pymc_samples.std())
# (6.647935041418325, 81.64505597965584)

# Shoud match this 
scipy_samples = st.halfcauchy(loc=0, scale=.01).rvs(10_000)
print(scipy_samples.mean(), scipy_samples.std())
# (0.049328030173446065, 0.3650001369075893)

# But matches this
scipy_samples = st.halfcauchy(loc=0.01, scale=1).rvs(10_000)
print(scipy_samples.mean(), scipy_samples.std())
# (5.537612320455185, 73.96359661595363)

# V3
pymc_samples = pm.HalfCauchy.dist(beta=.01, shape=10_000).random()      
print(pymc_samples.mean(), pymc_samples.std())                          
# (0.05829908302484347, 1.122931711761449)
```

The HalfCauchy is notoriously hard to evaluate, but with these parameters it seems reliable.

**Comment by brandonwillard** [comment]
> The random tests are still disabled, so this wouldn't show up right?

The tests in `pymc3.tests.test_distributions` perform some sampling and those are running for all the converted `Distribution`s; otherwise, I believe you're talking about `pymc3.tests.test_distributions_random`.  Those tests are *not* enabled because they're redundant (for the `RandomVariable`s that already exist in Aesara, at least).

>  I wanted to check with the prior predictive sampling but right now that's also unusable.

It's not unusable; you only need to set/change the seed between calls to `pm.sample_*_predictive`.  Regardless, I've pushed a commit that correctly sets `RandomVariable.inplace = True`, so you shouldn't need to do that anymore.

**Comment by ricardoV94** [comment]
Thanks Brandon. 

I didn't mean it's unusable. I just didn't know how to make use of it :b


I confirmed my suspicion about the Half Cauchy though. There is a mismatch between the old logp and the new randomOp, as the later expects two arguments, while the former expects only one. 

That's why the logp test would fail if the unused `alpha` argument was not added here: https://github.com/pymc-devs/pymc3/blob/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d/pymc3/distributions/continuous.py#L2276

It's not too difficult to add this loc parameter to the logp, but just wanted to make sure we want to do that. Do we want to adapt to the number of arguments expected in aesara or just customize our own random ops in terms of the arguments we want to use?

**Comment by brandonwillard** [comment]
> It's not too difficult to add this loc parameter to the logp, but just wanted to make sure we want to do that. Do we want to adapt to the number of arguments expected in aesara or just customize our own random ops in terms of the arguments we want to use?

If an adaptation is something as simple as inverting a parameter, then we'll always want to do that, but this case sounds like we'll need to do the latter (i.e. customize a `RandomVariable`).  There's now an [example of such a customization for `Multinomial`](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/multivariate.py#L434).

**Comment by ricardoV94** [comment]
On another front:

```python
import pymc3 as pm
with pm.Model() as m:
    x = pm.Uniform('x', 0, 1)
    y = pm.Uniform('y', 0, x)
    
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.83990034)
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.09721045)
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.23865389)
```

Also it ignores when I set transform=None, and creates the interval variables anyway.

**Comment by michaelosthege** [comment]
@ricardoV94 were the trailing `__` on the transformed variables removed?

On `master`:
```python
>>> m.logp({'x_interval': 0, 'y_interval': 0})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\osthege\Repos\pymc3-dev\pymc3\model.py", line 1565, in __call__
    return self.f(**point)
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3-dev\lib\site-packages\aesara\compile\function\types.py", line 956, in __call__
    raise TypeError(
TypeError: Missing required input: x_interval__ ~ TransformedDistribution
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
```

**Comment by ricardoV94** [comment]
Yes they were removed for the time being.

**Comment by brandonwillard** [comment]
> Also it ignores when I set transform=None, and creates the interval variables anyway.

I'll add fixes for those in a minute.

**Review by michaelosthege — COMMENTED** [review]
Made these comments a while back, but forgot to hit submit..

**Comment by ricardoV94** [comment]
@brandonwillard I poked around after you made the missing data changes to check whether we are auto-transforming the imputed variables or not, but I am finding some issues before even getting there. 

In V3:
```python
with pm.Model() as m:
  x = pm.Gamma('x', 1, 1, observed=np.array([1, 1, np.nan, 1]))
print(m.vars)  # [x_missing ~ NoDistribution]
print(m.logp({'x_missing': [1]}))  # 4.0
```
in V4:
```python
with pm.Model() as m:
  x = pm.Gamma('x', 1, 1, observed=np.array([1, 1, np.nan, 1]))
# Not sure if we should add `x` to m.vars...
print(m.vars)  # [ ]

# Evaluation is stochastic
print(m.logp({'x_missing': [1]}))  # -4.487674097730719
print(m.logp({'x_missing': [1]}))  # -3.195617080537621

# I can put any variable name, shouldn't this raise an Error regardless?
print(m.logp({'z': [1]}))  # -3.165731061987242
```

The random Gamma variable is still in the logp graph:
```python
print(aesara.dprint(m.logpt))
Sum{acc_dtype=float64} [id A] '__logp'   
 |MakeVector{dtype='float64'} [id B] ''   
   |Sum{acc_dtype=float64} [id C] ''   
     |Elemwise{mul,no_inplace} [id D] '__logp_x'   
       |Sum{acc_dtype=float64} [id E] ''   
       | |Elemwise{switch,no_inplace} [id F] ''   
       |   |Elemwise{mul,no_inplace} [id G] ''   
       |   | |Elemwise{mul,no_inplace} [id H] ''   
       |   | | |Elemwise{mul,no_inplace} [id I] ''   
       |   | | | |InplaceDimShuffle{x} [id J] ''   
       |   | | | | |TensorConstant{1} [id K]
       |   | | | |Elemwise{mul,no_inplace} [id L] ''   
       |   | | |   |InplaceDimShuffle{x} [id M] ''   
       |   | | |   | |TensorConstant{1} [id N]
       |   | | |   |Elemwise{ge,no_inplace} [id O] ''   
       |   | | |     |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   | | |     | |DisconnectedGrad [id Q] ''   
       |   | | |     | | |halfnormal_rv.1 [id R] 'x'   
       |   | | |     | |   |RandomStateSharedVariable(<RandomState(MT19937) at 0x7F4779136140>) [id S]
       |   | | |     | |   |TensorConstant{(1,) of 4} [id T]
       |   | | |     | |   |TensorConstant{11} [id U]
       |   | | |     | |   |TensorConstant{1.0} [id V]
       |   | | |     | |   |Elemwise{true_div,no_inplace} [id W] ''   
       |   | | |     | |     |TensorConstant{1.0} [id X]
       |   | | |     | |     |Elemwise{inv,no_inplace} [id Y] ''   
       |   | | |     | |       |TensorConstant{1.0} [id Z]
       |   | | |     | |TensorConstant{(3,) of 1.0} [id BA]
       |   | | |     | |TensorConstant{[ True  Tr..lse  True]} [id BB]
       |   | | |     |InplaceDimShuffle{x} [id BC] ''   
       |   | | |       |TensorConstant{0} [id BD]
       |   | | |InplaceDimShuffle{x} [id BE] ''   
       |   | |   |Elemwise{mul,no_inplace} [id BF] ''   
       |   | |     |TensorConstant{1} [id BG]
       |   | |     |Elemwise{gt,no_inplace} [id BH] ''   
       |   | |       |TensorConstant{1.0} [id V]
       |   | |       |TensorConstant{0} [id BI]
       |   | |InplaceDimShuffle{x} [id BJ] ''   
       |   |   |Elemwise{mul,no_inplace} [id BK] ''   
       |   |     |TensorConstant{1} [id BL]
       |   |     |Elemwise{gt,no_inplace} [id BM] ''   
       |   |       |Elemwise{true_div,no_inplace} [id BN] ''   
       |   |       | |TensorConstant{1.0} [id X]
       |   |       | |Elemwise{inv,no_inplace} [id BO] ''   
       |   |       |   |TensorConstant{1.0} [id Z]
       |   |       |TensorConstant{0} [id BP]
       |   |Elemwise{add,no_inplace} [id BQ] ''   
       |   | |Elemwise{sub,no_inplace} [id BR] ''   
       |   | | |InplaceDimShuffle{x} [id BS] ''   
       |   | | | |Elemwise{add,no_inplace} [id BT] ''   
       |   | | |   |Elemwise{neg,no_inplace} [id BU] ''   
       |   | | |   | |gammaln [id BV] ''   
       |   | | |   |   |TensorConstant{1.0} [id V]
       |   | | |   |Elemwise{switch,no_inplace} [id BW] ''   
       |   | | |     |Elemwise{eq,no_inplace} [id BX] ''   
       |   | | |     | |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | | |     | |TensorConstant{0} [id BY]
       |   | | |     |Elemwise{switch,no_inplace} [id BZ] ''   
       |   | | |     | |Elemwise{eq,no_inplace} [id CA] ''   
       |   | | |     | | |TensorConstant{1.0} [id V]
       |   | | |     | | |TensorConstant{0} [id CB]
       |   | | |     | |TensorConstant{0.0} [id CC]
       |   | | |     | |TensorConstant{-inf} [id CD]
       |   | | |     |Elemwise{mul,no_inplace} [id CE] ''   
       |   | | |       |TensorConstant{1.0} [id V]
       |   | | |       |Elemwise{log,no_inplace} [id CF] ''   
       |   | | |         |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | | |Elemwise{mul,no_inplace} [id CG] ''   
       |   | |   |InplaceDimShuffle{x} [id CH] ''   
       |   | |   | |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | |   |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   | |Elemwise{switch,no_inplace} [id CI] ''   
       |   |   |Elemwise{eq,no_inplace} [id CJ] ''   
       |   |   | |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   |   | |InplaceDimShuffle{x} [id CK] ''   
       |   |   |   |TensorConstant{0} [id CL]
       |   |   |InplaceDimShuffle{x} [id CM] ''   
       |   |   | |Elemwise{switch,no_inplace} [id CN] ''   
       |   |   |   |Elemwise{eq,no_inplace} [id CO] ''   
       |   |   |   | |Elemwise{sub,no_inplace} [id CP] ''   
       |   |   |   | | |TensorConstant{1.0} [id V]
       |   |   |   | | |TensorConstant{1} [id CQ]
       |   |   |   | |TensorConstant{0} [id CR]
       |   |   |   |TensorConstant{0.0} [id CS]
       |   |   |   |TensorConstant{-inf} [id CT]
       |   |   |Elemwise{mul,no_inplace} [id CU] ''   
       |   |     |InplaceDimShuffle{x} [id CV] ''   
       |   |     | |Elemwise{sub,no_inplace} [id CP] ''   
       |   |     |Elemwise{log,no_inplace} [id CW] ''   
       |   |       |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   |InplaceDimShuffle{x} [id CX] ''   
       |     |TensorConstant{-inf} [id CY]
       |TensorConstant{1.0} [id CZ]
```

**Comment by brandonwillard** [comment]
> @brandonwillard I poked around after you made the missing data changes to check whether we are auto-transforming the imputed variables or not, but I am finding some issues before even getting there.

Yes, exactly, there are no model-level random variables added to the graph for missing values.  The missing data look just like normal data to the model, and they don't need to be transformed.  Unless we're going to compute gradients in those missing terms and attempt to estimate them, why would we want to transform them.

Also, the evaluation should necessarily be stochastic if we're imputing missing values via sampling.  We might need to make sure that the RNG states are cloned between multiple sampler steps, though, or that the same `*Subtensor*`-updated observation array is used by all of them.

**Comment by ricardoV94** [comment]
I think we indeed used to infer these during sampling, using NUTS as usual. Wasn't that the point of having them?

**Comment by twiecki** [comment]
> I think we indeed used to infer these during sampling, using NUTS as usual. Wasn't that the point of having them?

Yes, that's right.

**Comment by ricardoV94** [comment]
Okay I was not aware that was a feature and not a bug. Are we sure we want, by default, to impute values by random sampling instead of NUTS as before?

This also means we can only impute values for distributions with random methods.

**Comment by brandonwillard** [comment]
Yes, for variables without a sampling method we're going to need to add them to the model to be sampled themselves.

I believe we can do that just as easily by changing the `RandomVariable` part of the `*IncSubtensor*` to be the array of unobserved terms instead, and simply add those terms to the model variables.

**Comment by ricardoV94** [comment]
Alright so the question then is what to use as the default, even if both ways are possible right?

**Comment by brandonwillard** [comment]
> Alright so the question then is what to use as the default, even if both ways are possible right?

Sampling the current way&mdash;i.e. some form of "exact" sampling&mdash;will be a lot cheaper, more "accurate", and faster in some cases, but definitely not all.

**Comment by brandonwillard** [comment]
N.B.: You know, this missing data situation is one of those cases in which having one Aesara graph that performs all the sampling steps would be a real advantage, because it would remove all the out-of-graph work required to sync and/or recompute values (especially exactly sampled missing data) across different sampler steps.  

For that matter, there could easily be entire (sub)graphs that we're currently recomputing between steps.  Also, performing in-place updates between the same steps is so much more difficult with the current approach, because we have to implement it by hand, whereas such things could be automatically applied as a single graph&mdash;even with the existing optimizations!

**Comment by ricardoV94** [comment]
Less important than this, from my comment above shouldn't the `model.logp` have raised an error when passing a dictionary with a variable not in the graph?

**Comment by brandonwillard** [comment]
> Less important than this, from my comment above shouldn't the `model.logp` have raised an error when passing a dictionary with a variable not in the graph?

No, no reason to do that.  Plus, the implementation would become unnecessarily more difficult that way.

**Comment by ricardoV94** [comment]
Sampling is failing when passing a dictionary for starting point with values for "untransformed" variables. The old test `TestDisasterModel` in `test_examples.py` that was marked as xfail shows this well:

```python
@pytest.mark.xfail(
    reason="_check_start_shape fails with start dictionary"
    # condition=(aesara.config.floatX == "float32"), reason="Fails on float32"
)
class TestDisasterModel(SeededTest):
    @pytest.mark.xfail(condition=(aesara.config.floatX == "float32"), reason="Fails on float32")
    # Time series of recorded coal mining disasters in the UK from 1851 to 1962
    def test_disaster_model(self):
        model = build_disaster_model(masked=False)
        with model:
            # Initial values for stochastic nodes
            start = {"early_mean": 2, "late_mean": 3.0}
            # Use slice sampler for means (other variables auto-selected)
            step = pm.Slice([model["early_mean_log__"], model["late_mean_log__"]])
            tr = pm.sample(500, tune=50, start=start, step=step, chains=2)
```
The problem arises in the function `_check_start_shape` which passes the whole `start` dictionary (including the two original key-value pairs) leading to an exception on aesara side because those variables are not valid inputs. 

I was able to run this on my branch where I have refactored already the `DiscreteUniform`, which was the reason why this text was marked as an xfail, but it should be straightforward to get an example that is compatible with the current V4 branch.

**Comment by brandonwillard** [comment]
> Sampling is failing when passing a dictionary for starting point with values for "untransformed" variables. The old test `TestDisasterModel` in `test_examples.py` that was marked as xfail shows this well:

So we can update the test to pass the transformed variables, or use/implement a function that converts a point `dict` with untransformed variables to transformed ones within `pm.sample`, no?

**Comment by ricardoV94** [comment]
> So we can update the test to pass the transformed variables, or use/implement a function that converts a point `dict` with untransformed variables to transformed ones within `pm.sample`, no?

The `update_start_vals` function already converts transformed start vals to untransformed ones, so we just need a way to filter the original transformed values after this is done.

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**